store/sql_team_store.go

```diff
@@ -583,29 +583,63 @@ func (s SqlTeamStore) GetTeamsForUser(userId string) StoreChannel {
 	return storeChannel
 }
 
+type channelsUnread struct {
+	TeamId        string
+	TotalMsgCount int64
+	MsgCount      int64
+	MentionCount  int64
+	NotifyProps   model.StringMap
+}
+
 func (s SqlTeamStore) GetTeamsUnreadForUser(teamId, userId string) StoreChannel {
 	storeChannel := make(StoreChannel, 1)
 
 	go func() {
 		result := StoreResult{}
 
-		var members []*model.TeamUnread
-		_, err := s.GetReplica().Select(&members,
+		var data []channelsUnread
+		_, err := s.GetReplica().Select(&data,
 			`SELECT
-				Channels.TeamId,
-				SUM(Channels.TotalMsgCount - ChannelMembers.MsgCount) as MsgCount,
-				SUM(ChannelMembers.MentionCount) as MentionCount
+				Channels.TeamId, Channels.TotalMsgCount, ChannelMembers.MsgCount, ChannelMembers.MentionCount, ChannelMembers.NotifyProps
 			FROM
-				Channels,
-				ChannelMembers
+				Channels, ChannelMembers
 			WHERE
-				Channels.Id = ChannelMembers.ChannelId AND Channels.DeleteAt = 0 AND
-				ChannelMembers.UserId = :UserId AND Channels.TeamId != :TeamId
-			GROUP BY
-				Channels.TeamId`, map[string]interface{}{"UserId": userId, "TeamId": teamId})
+				Id = ChannelId AND UserId = :UserId AND DeleteAt = 0 AND TeamId != :TeamId`,
+			map[string]interface{}{"UserId": userId, "TeamId": teamId})
+
 		if err != nil {
 			result.Err = model.NewLocAppError("SqlTeamStore.GetTeamsUnreadForUser", "store.sql_team.get_unread.app_error", nil, "userId="+userId+" "+err.Error())
 		} else {
+			var members []*model.TeamUnread
+			membersMap := make(map[string]*model.TeamUnread)
+
+			unreads := func(cu channelsUnread, tu *model.TeamUnread) *model.TeamUnread {
+				tu.MentionCount += cu.MentionCount
+
+				if cu.NotifyProps["mark_unread"] != model.CHANNEL_MARK_UNREAD_MENTION {
+					tu.MsgCount += (cu.TotalMsgCount - cu.MsgCount)
+				}
+
+				return tu
+			}
+
+			for i := range data {
+				id := data[i].TeamId
+				if mu, ok := membersMap[id]; ok {
+					membersMap[id] = unreads(data[i], mu)
+				} else {
+					membersMap[id] = unreads(data[i], &model.TeamUnread{
+						MsgCount:     0,
+						MentionCount: 0,
+						TeamId:       id,
+					})
+				}
+			}
+
+			for _, val := range membersMap {
+				members = append(members, val)
+			}
+
```

Comment:
This logic doesn't really belong in the store, can you move it out to the api package?